### PR TITLE
Search function

### PR DESCRIPTION
**Pull request by dthanhtuan**

### CODE REVIEW

File: app/forms/developers_search_form.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  attribute :prog_lang, String
+  attribute :language_code, String
+  attribute :show_all, Boolean
+
+  def search
+    if show_all
+      query = Developer.includes(:programming_languages, :languages)
+    elsif !email.present? && !prog_lang.present? && !language_code.present?
+      Developer.none
+    else
```
Comment (thienhv) [inline app/forms/developers_search_form.rb]: @dthanhtuan 
I think you should use Ruby style at here

```
return `all` if show_all
return `none` if `xxx`
do another case
```

File: app/forms/developers_search_form.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      query = Developer.includes(:programming_languages, :languages, :developer_programming_languages, :developer_languages)
+      query = query.where('developers.email like :search_string', search_string: "%#{sanitize_sql_like(email)}%") if email.present?
```
Comment (thienhv) [inline app/forms/developers_search_form.rb]: @dthanhtuan 

screen shot 2018-04-04 at 14 28 54

File: spec/factories.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+FactoryBot.define do
+
+  factory :developer, class: Developer do
+    email 'mrabc@gmail.com'
```
Comment (thienhv) [inline spec/factories.rb]: @dthanhtuan 
use `Faker` to create dummy data

https://github.com/stympy/faker#contents

File: spec/forms/developers_search_form_spec.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
+require 'rails_helper'
+RSpec.describe DevelopersSearchForm do
+  describe '#search' do
+    before do
+      programming_language_php = create :programming_language, name: 'PHP'
+      programming_language_ruby = create :programming_language, name: 'Ruby'
+      language_code_en = create :language, code: 'en'
+      create :developer, email: 'mrtest@gmail.com', programming_languages: [programming_language_php], languages: [language_code_en]
+      create :developer, email: 'tuan.dao@tech.reapra.sg', programming_languages: [programming_language_ruby], languages: [language_code_en]
```
Comment (thienhv) [inline spec/forms/developers_search_form_spec.rb]: @dthanhtuan 
l5,l6 you should remove it
and change l8,l9
ex:
```
create :developer, email: 'mrtest@gmail.com', programming_languages: [create(:programming_language, name: 'PHP')], languages: [language_code_en]
```

File: spec/system/developer_search_spec.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    language_code_en = create :language, code: 'en'
+    create :developer, email: 'mrtest@gmail.com', programming_languages: [programming_language_php], languages: [language_code_en]
+    create :developer, email: 'tuan.dao@tech.reapra.sg', programming_languages: [programming_language_ruby], languages: [language_code_en]
+  end
+
+  it 'Search with empty params' do
+    visit developers_path
+    click_on 'Search'
+    expect(page).to have_content('No result')
+    sleep 2
```
[the remark this answers is not in the document]
Comment (dthanhtuan) [reply spec/system/developer_search_spec.rb]: I want to watch the test running steps, will remove it :)

File: app/forms/developers_search_form.rb (hunk 2 above)
Comment (dthanhtuan) [reply app/forms/developers_search_form.rb]: I have said this to ryuji-san and he allow me to keep email attribute in developers search form.
> should I remove it ? Search query still works without email input
> Then it's fine. But can you also tell that to thienhv when asking him for review?

File: spec/forms/developers_search_form_spec.rb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+RSpec.describe DevelopersSearchForm do
+  describe '#search' do
+    before do
+      language_code_en = create :language, code: 'en'
+      create :developer, email: 'mrtest@gmail.com', programming_languages: [create(:programming_language, name: 'PHP')], languages: [language_code_en]
+      create :developer, email: 'tuan.dao@tech.reapra.sg', programming_languages: [create(:programming_language, name: 'Ruby')], languages: [language_code_en]
+    end
+
+    context 'with no params' do
+      it 'shows empty lít' do
```
[the remark this answers is not in the document]
Comment (dthanhtuan) [reply spec/forms/developers_search_form_spec.rb]: this is 'list', I accidentally type list when opening an unicode program which uses for typing vietnamese.
Sorry about that :)

**Pull request merged**